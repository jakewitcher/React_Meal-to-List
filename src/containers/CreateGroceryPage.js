import React, { Component } from 'react';
import { connect } from 'react-redux';
import GroceryForm from '../components/grocery/GroceryForm';
import GroceryItems from '../components/grocery/GroceryItems';

class CreateGroceryPage extends Component {
    state = {
        groceryName: "",
        selectedMeal: this.props.meals.mealList[0] || {},
        itemMap: new Map(),
    };

    addMealToMap = () => {
        const { itemMap, selectedMeal } = this.state;
        const CombinedItemsList = selectedMeal.itemList.reduce((a, b) => {
            if (a.has(b.itemName)) {
                const newAmount = Number(a.get(b.itemName).amount) + Number(b.amount);
                const itemCopy = Object.assign({}, b)
                itemCopy.amount = newAmount;
                a.set(itemCopy.itemName, itemCopy);
                return a;
            }
            a.set(b.itemName, b);
            return a;
        }, itemMap);
        this.setState({
            itemMap: CombinedItemsList,
            selectedMeal: this.props.meals.mealList[0],
        })
    };

    selectedMealChange = (e) => {
        const { meals } = this.props;
        const selectedMealName = e.target.value;
        const selectedMeal = meals.mealList.filter(meal => meal.name === selectedMealName);
        this.setState({
            selectedMeal: selectedMeal[0],
        });
    };

    groceryNameChange = (e) => {
        const groceryName = e.target.value;
        this.setState({ groceryName });
    };

    render() {
        return (
            <div className="grocery">
                <GroceryForm 
                    meals={this.props.meals}
                    selectedMeal={this.state.selectedMeal}
                    groceryName={this.state.groceryName}
                    selectedMealChange={this.selectedMealChange}
                    groceryNameChange={this.groceryNameChange}
                    addMealToMap={this.addMealToMap}
                />
                <GroceryItems 
                    groceryName={this.state.groceryName}
                />
            </div>
        )
    }

};

const mapStateToProps = (state) => {
    return {
        meals: state.meals,
    }
};

export default connect(mapStateToProps)(CreateGroceryPage);