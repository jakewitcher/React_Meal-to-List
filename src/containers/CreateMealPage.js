import React from 'react';
import CreateMealForm from '../components/meal/CreateMealForm';
import MealItems from '../components/meal/MealItems';
import { addMeal } from '../actions/meal';


const CreateMealPage = (props) => {
    return (
        <div>
            <div className="meal">
                <CreateMealForm 
                    {...props}
                    addMeal={addMeal} 
                />
                <MealItems
                    mealName={props.mealProps.mealName}
                    itemList={props.mealProps.itemList}
                />
            </div>
        </div>
       
    )
};

export default CreateMealPage;