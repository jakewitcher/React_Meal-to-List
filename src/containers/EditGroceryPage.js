import React from 'react';

const EditGroceryPage = (props) => {
    return (
        <div>
            <h1>This is the Edit Grocery Page</h1>
        </div>
    );
}

export default EditGroceryPage;