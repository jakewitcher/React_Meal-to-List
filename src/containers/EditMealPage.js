import React from 'react';

const EditMealPage = (props) => {
    return (
        <div>
            <h1>This is the Edit Meal Page</h1>
        </div>
    );
}

export default EditMealPage;