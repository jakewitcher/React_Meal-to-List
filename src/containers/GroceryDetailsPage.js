import React from "react";
import { connect } from "react-redux";
import GroceryItems from "../components/grocery/GroceryItems";

export const GroceryDetailsPage = props => {
  return (
    <div className="details__container">
      <div className="details">
        <GroceryItems
          groceryName={props.grocery.name}
          itemList={props.grocery.itemList}
          isDetails={true}
        />
      </div>
    </div>
  );
};

const mapStateToProps = (state, props) => {
  return {
    groceryLists: state.groceryLists,
    grocery: state.groceryLists.groceryList.filter(
      list => list.id === props.match.params.groceryId
    )[0]
  };
};

export default connect(mapStateToProps)(GroceryDetailsPage);
