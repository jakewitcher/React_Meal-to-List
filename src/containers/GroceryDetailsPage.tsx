import React from "react";
import { connect } from "react-redux";
import GroceryItems from "../components/GroceryItems";
import { ApplicationState } from "../store/index";
import { RouteComponentProps } from "react-router";
import { Grocery } from "../store/grocery/types";

interface IGroceryDetailsPageProps
  extends RouteComponentProps<{ groceryId: string }> {
  grocery: Grocery;
}

export const GroceryDetailsPage: React.FC<IGroceryDetailsPageProps> = props => {
  return (
    <div className="details__container">
      <div className="details">
        <GroceryItems
          groceryName={props.grocery.name}
          itemList={props.grocery.itemList}
          isDetails={true}
        />
      </div>
    </div>
  );
};

const mapStateToProps = (
  state: ApplicationState,
  props: IGroceryDetailsPageProps
) => {
  return {
    groceryLists: state.groceryLists,
    grocery: state.groceryLists.groceryList.filter(
      list => list.id === props.match.params.groceryId
    )[0]
  };
};

export default connect(mapStateToProps)(GroceryDetailsPage);
