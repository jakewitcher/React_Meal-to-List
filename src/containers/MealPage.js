import React, { Component } from 'react';
import { connect } from 'react-redux';

class MealPage extends Component {
    render () {
        return (
            <div>
                <h1>Meal Page</h1>
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        meals: state.meals
    }
} 

export default connect(mapStateToProps)(MealPage);