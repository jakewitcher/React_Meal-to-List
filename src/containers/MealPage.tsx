import React, { Component, ChangeEvent } from "react";
import { Dispatch } from "redux";
import { connect } from "react-redux";
import MealForm from "../components/MealForm";
import MealItems from "../components/MealItems";
import { onAddMeal, onEditMeal } from "../store/meal/actions";
import { onAddItem } from "../store/item/actions";
import { Item, ItemState } from "../store/item/types";
import { Meal } from "../store/meal/types";
import { ApplicationState } from "../store/index";
import { RouteComponentProps } from "react-router";

interface IMealPageProps extends RouteComponentProps<{ mealId: string }> {
  meal?: Meal;
  items: ItemState;
  onAddItem(name: string): void;
  onEditMeal(meal: Meal): void;
  onAddMeal(meal: Meal): void;
}

interface IMealPageState {
  mealName: string;
  itemName: string;
  amount: string;
  unit: string;
  itemList: Item[];
}

export class MealPage extends Component<IMealPageProps, IMealPageState> {
  constructor(props: IMealPageProps) {
    super(props);

    this.state = {
      mealName: props.meal ? props.meal.name : "",
      itemName: "",
      amount: "",
      unit: "",
      itemList: props.meal ? props.meal.itemList : []
    };
  }

  deleteMealItem = (name: string) => {
    const updateMealList = this.state.itemList.filter(
      item => item.itemName !== name
    );
    this.setState({
      itemList: updateMealList
    });
  };

  itemToEdit = (name: string) => {
    const selectedItem = this.state.itemList.filter(
      item => item.itemName === name
    )[0];
    this.setState({
      itemName: selectedItem.itemName,
      amount: selectedItem.amount.toString(),
      unit: selectedItem.unit
    });
  };

  mealNameChange = (e: ChangeEvent<HTMLInputElement>) => {
    const mealName = e.target.value;
    this.setState({ mealName });
  };

  itemNameChange = (e: ChangeEvent<HTMLInputElement>) => {
    const itemName = e.target.value;
    this.setState({ itemName });
  };

  amountChange = (e: ChangeEvent<HTMLInputElement>) => {
    const amount = e.target.value;
    this.setState({ amount });
  };

  unitChange = (e: ChangeEvent<HTMLSelectElement>) => {
    const unit = e.target.value;
    this.setState({ unit });
  };

  findItem = (itemName: string, itemList: string[]) => {
    if (itemList.find(item => item === itemName)) {
      return true;
    }
    return false;
  };

  // checks if item is in the master list of items. If not, the item is added.
  addToItemsAll = (items: ItemState, itemName: string) => {
    if (!this.findItem(itemName, items.itemsAll.map(i => i.name))) {
      this.props.onAddItem(itemName);
    }
  };

  // checks if the item is already in the meal list. If it is, the current item is replaced with the udpated item information.
  addToItems = (itemList: Item[], itemName: string, newItem: Item) => {
    if (this.findItem(itemName, itemList.map(i => i.itemName))) {
      itemList = itemList.filter(item => item.itemName !== itemName);
    }

    this.setState({
      itemName: "",
      amount: "",
      itemList: [...itemList, newItem]
    });
  };

  addItem = () => {
    let { itemName, amount, unit, itemList } = this.state;
    const { items } = this.props;
    let amountToNum = parseFloat(amount);
    const newItem = {
      itemName,
      amount: amountToNum,
      unit
    };

    this.addToItemsAll(items, itemName);
    this.addToItems(itemList, itemName, newItem);
  };

  resetMeal = () => {
    this.setState({ mealName: "", itemList: [] });
  };

  render() {
    return (
      <div className="update">
        <MealForm
          mealNameChange={this.mealNameChange}
          itemNameChange={this.itemNameChange}
          amountChange={this.amountChange}
          unitChange={this.unitChange}
          addItem={this.addItem}
          resetMeal={this.resetMeal}
          mealName={this.state.mealName}
          itemName={this.state.itemName}
          amount={this.state.amount}
          unit={this.state.unit}
          itemList={this.state.itemList}
          mealId={this.props.meal ? this.props.meal.id : ""}
          updateMeal={
            this.props.meal ? this.props.onEditMeal : this.props.onAddMeal
          }
        />
        <MealItems
          mealName={this.state.mealName}
          itemList={this.state.itemList}
          itemToEdit={this.itemToEdit}
          deleteMealItem={this.deleteMealItem}
          isDetails={false}
        />
      </div>
    );
  }
}

const mapStateToProps = (state: ApplicationState, props: IMealPageProps) => {
  return {
    meals: state.meals,
    items: state.items,
    meal: state.meals.mealList.filter(
      meal => meal.id === props.match.params.mealId
    )[0]
  };
};

const mapDispatchToProps = (dispatch: Dispatch) => {
  return {
    onAddItem: (data: string) => dispatch(onAddItem(data)),
    onEditMeal: (data: Meal) => dispatch(onEditMeal(data)),
    onAddMeal: (data: Meal) => dispatch(onAddMeal(data))
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MealPage);
