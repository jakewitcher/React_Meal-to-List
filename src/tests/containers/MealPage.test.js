import React from "react";
import { shallow } from "enzyme";
import { MealPage } from "../../containers/MealPage";

test("should render MealPage correctly", () => {
  const wrapper = shallow(<MealPage />);
  expect(wrapper).toMatchSnapshot();
});
