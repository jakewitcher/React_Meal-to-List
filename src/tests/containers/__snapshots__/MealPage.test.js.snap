// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MealPage correctly 1`] = `
<div
  className="meal"
>
  <MealForm
    addItem={[Function]}
    amount=""
    amountChange={[Function]}
    itemList={Array []}
    itemName=""
    itemNameChange={[Function]}
    mealId=""
    mealName=""
    mealNameChange={[Function]}
    resetMeal={[Function]}
    title="Create Meal"
    unit="pound(s)"
    unitChange={[Function]}
  />
  <MealItems
    deleteMealItem={[Function]}
    itemList={Array []}
    itemToEdit={[Function]}
    mealName=""
  />
</div>
`;
